VMEncryption/main/patch/UbuntuPatching.py

```diff
@@ -47,33 +47,44 @@ def __init__(self, logger, distro_info):
         self.umount_path = '/bin/umount'
         self.touch_path = '/usr/bin/touch'
 
+    def packages_installed(self, packages):
+        ''' return true if all packages in list are already installed '''
+        installed = True
+        for package in packages:
+            cmd = "dpkg-query -s {package} | grep -q 'install ok installed'"
+            if not self.command_executor.ExecuteInBash(cmd, False, None, None, True):
+                installed = False
+                self.logger.log("{1} package not yet installed".format(package))
+        return installed
+
     def install_cryptsetup(self):
         packages = ['cryptsetup-bin']
-        cmd = " ".join(['apt-get', 'install', '-y', '--no-upgrade'] + packages)
-        return_code = self.command_executor.Execute(cmd, timeout=30)
-        if return_code == -9:
-            msg = "Command: apt-get install timed out. Make sure apt-get is configured correctly and there are no network problems."
-            raise Exception(msg)
-
-        # If install fails, try running apt-get update and then try install again
-        if return_code != 0:
-            self.logger.log('cryptsetup installation failed. Retrying installation after running update')
-            return_code = self.command_executor.Execute('apt-get -o Acquire::ForceIPv4=true -y update', timeout=30)
-            # Fail early if apt-get update times out.
-            if return_code == -9:
-                msg = "Command: apt-get -o Acquire::ForceIPv4=true -y update timed out. Make sure apt-get is configured correctly."
-                raise Exception(msg)
-            cmd = " ".join(['apt-get', 'install', '-y'] + packages)
+
+        if self.packages_installed(packages):
+            return
+        else:
+            cmd = " ".join(['apt-get', 'install', '-y', '--no-upgrade'] + packages)
```

Comment:
I thought --no-upgrade will not download if the package is already installed. Is that not working as expected.
See the --no-upgrade option description: https://linux.die.net/man/8/apt-get

Reply:
The issue here is that we are trying to prevent an outbound call, which can timeout on isolated networks, and then cause an extension timeout.